### PR TITLE
Implement Structure.dump_cmd/3

### PR DESCRIPTION
**Pull request by maennchen**
See https://github.com/elixir-ecto/ecto_sql/pull/423

* [x] TODO: Change `mix.exs` back to released versions.

### CODE REVIEW

**Comment by maennchen** [comment]
I think the CI failure is not caused by this change.

**Comment by warmwaffles** [comment]
@maennchen try updating / rebasing this branch off of `main`. This was recently merged https://github.com/elixir-sqlite/ecto_sqlite3/pull/80

**Comment by greg-rychlewski** [comment]
@warmwaffles I think I see one of the other issues. There was an `insert_all` integration test added that uses select statements for values and it doesn't seem like your adapter supports that. Which tag do you use to filter those? I can make the change in Ecto.

**Comment by greg-rychlewski** [comment]
Another one was because stacktraces were added to logging and the test didn't take into account that for adapters like yours the integration tests are under the `deps` directory. Fixing it here: https://github.com/elixir-ecto/ecto_sql/pull/426.

There is also one other failure that looks like it has something to do with this change: https://github.com/elixir-ecto/ecto_sql/pull/399. But haven't figured it out yet.

**Comment by maennchen** [comment]
@warmwaffles I've rebased onto `main`. There's however still a few problems left as @greg-rychlewski noticed.

**Comment by warmwaffles** [comment]
> Probably number 1 is better, because this is a super specific thing for sqlite? But let me know if you have a different idea.

This is probably a better idea to do.

It's probably okay to merge this into `main` and then follow up with tests. I am unsure why this PR doesn't have github actions running for it.

**Comment by warmwaffles** [comment]
I haven't forgotten about this, just haven't had time to dig in more.

**Comment by greg-rychlewski** [comment]
I think the actions are only running the first time the PR is opened because of this

```
pull_request:
    types:
      - opened
```

I think you need the `synchronize` activity for it to trigger on new commits.

**Comment by warmwaffles** [comment]
@maennchen would you mind rebasing off of the latest `main` branch? I want to see if the workflow change I made gets picked up and kicks off.

**Comment by maennchen** [comment]
@warmwaffles Rebased. Tests seem to be running :+1:

**Comment by warmwaffles** [comment]
Excellent, thanks @greg-rychlewski for pointing out that I had that `types` section.

**Comment by greg-rychlewski** [comment]
btw failed integration tests 1 and 3 can be fixed by updating to the latest commit on ecto/ecto_sql:

https://github.com/elixir-ecto/ecto/pull/3942
https://github.com/elixir-ecto/ecto_sql/pull/426

**Comment by maennchen** [comment]
@greg-rychlewski Updated

**Comment by greg-rychlewski** [comment]
@maennchen I sent a PR to your fork to fix the broken unit tests. the planner now returns the cast and dump version of the parameters and it was causing a match error.

**Comment by greg-rychlewski** [comment]
@maennchen if you rebase on master the tests should all pass now

**Comment by maennchen** [comment]
@greg-rychlewski Still seems to error, unfortunately.

**Comment by greg-rychlewski** [comment]
@maennchen oh it looks like the PR I sent you was reverted: https://github.com/maennchen/ecto_sqlite3/pull/1/files. if that change is re-implemented it should all pass

**Comment by maennchen** [comment]
@greg-rychlewski Oops, I rebased away your commit. I've brought it back.

File: mix.exs [hunk 1]
```diff
@@ -34,8 +34,9 @@ defmodule EctoSQLite3.MixProject do
   defp deps do
     [
       {:decimal, "~> 1.6 or ~> 2.0"},
-      {:ecto_sql, "~> 3.7"},
-      {:ecto, "~> 3.7"},
+      # TODO: Change back to released version
+      {:ecto_sql, "~> 3.9.0-dev", github: "elixir-ecto/ecto_sql"},
+      {:ecto, "~> 3.9.0-dev", github: "elixir-ecto/ecto"},
```
Comment (warmwaffles) [inline mix.exs:39]: Would we be able to get these back to official packages?

Comment (maennchen) [reply mix.exs:39]: @warmwaffles there was no release yet.

Comment (maennchen) [reply mix.exs]: @warmwaffles The packages are available. I changed the `mix.exs` and updated.

**Comment by maennchen** [comment]
After the update, the following test breaks:

```
== Compilation error in file integration_test/json_test.exs ==
** (KeyError) key :meta not found
    expanding struct: Ecto.Integration.Order.__struct__/1
    integration_test/json_test.exs:10: Ecto.Integration.JsonTest."test json_extract_path with primitive values"/1
```

**Comment by maennchen** [comment]
After the fix 4 tests still fail with similar errors:

```
  * test query select selected_as/2 with having (3.9ms) [L#1608]

  1) test query select selected_as/2 with having (Ecto.Integration.RepoTest)
     deps/ecto/integration_test/cases/repo.exs:1608
     ** (Ecto.QueryError) unsupported expression :min_visits in query:
     
     from p0 in Ecto.Integration.Post,
       group_by: [p0.posted],
       having: selected_as(:min_visits) > 0,
       or_having: not (selected_as(:min_visits) > 0),
       order_by: [asc: p0.posted],
       select: %{posted: p0.posted, min_visits: selected_as(min(coalesce(p0.visits, 0)), :min_visits)}
     
     code: |> TestRepo.all()
     stacktrace:
       (ecto_sqlite3 0.8.1) lib/ecto/adapters/sqlite3/connection.ex:1403: Ecto.Adapters.SQLite3.Connection.expr/3
       (ecto_sqlite3 0.8.1) lib/ecto/adapters/sqlite3/connection.ex:1806: Ecto.Adapters.SQLite3.Connection.intersperse_map/4
       (ecto_sqlite3 0.8.1) lib/ecto/adapters/sqlite3/connection.ex:1356: Ecto.Adapters.SQLite3.Connection.expr/3
       (ecto_sqlite3 0.8.1) lib/ecto/adapters/sqlite3/connection.ex:1353: Ecto.Adapters.SQLite3.Connection.expr/3
       (ecto_sqlite3 0.8.1) lib/ecto/adapters/sqlite3/connection.ex:1172: Ecto.Adapters.SQLite3.Connection.paren_expr/3
       (ecto_sqlite3 0.8.1) lib/ecto/adapters/sqlite3/connection.ex:1143: Ecto.Adapters.SQLite3.Connection.boolean/4
       (ecto_sqlite3 0.8.1) lib/ecto/adapters/sqlite3/connection.ex:171: Ecto.Adapters.SQLite3.Connection.all/2
       (ecto_sqlite3 0.8.1) lib/ecto/adapters/sqlite3.ex:176: Ecto.Adapters.SQLite3.prepare/2
       (ecto 3.9.0) lib/ecto/query/planner.ex:182: Ecto.Query.Planner.query_without_cache/4
       (ecto 3.9.0) lib/ecto/query/planner.ex:152: Ecto.Query.Planner.query_prepare/6
       (ecto 3.9.0) lib/ecto/query/planner.ex:127: Ecto.Query.Planner.query_with_cache/8
       (ecto 3.9.0) lib/ecto/repo/queryable.ex:211: Ecto.Repo.Queryable.execute/4
       (ecto 3.9.0) lib/ecto/repo/queryable.ex:19: Ecto.Repo.Queryable.all/3
       deps/ecto/integration_test/cases/repo.exs:1624: (test)

```

**Comment by greg-rychlewski** [comment]
@maennchen You can add these tags to [the ignore list for the integration tests](https://github.com/elixir-sqlite/ecto_sqlite3/blob/main/integration_test/test_helper.exs#L74):

:selected_as_with_group_by
:selected_as_with_order_by
:selected_as_with_order_by_expression
:selected_as_with_having

**Comment by warmwaffles** [comment]
@maennchen I believe if you drop the changes made to the `mix.lock` the conflict should be resolved.

**Comment by maennchen** [comment]
@warmwaffles I've rebased and it seems the tests are failing again.

I've spent more time now rebasing this PR than writing it. I would welcome it if a maintainer makes sure that this PR is incorporated in a way that actually works.

**Comment by warmwaffles** [comment]
@maennchen I'll see what I can do and try to retain your original authorship. Definitely want to keep your contributions attached.

**Comment by warmwaffles** [comment]
Thanks a ton @maennchen for implementing this. It's been merged and deployed under `v0.9.0`

**Pull request closed without merging**